degree(multivariate): multivariate expression now requires generator to specified

akash9712 opened the pull request:
multivariate expressions could be passed into the degree expression, in
which case the degree was returned for the first generator. As of this pull
request, when giving the degree function multivariate expressions, a
generator is mandatory to be specified, else a TypeError is raised.

Example:
```
>>> degree(x+x**2)
2
>>>degree(x**2+y**3,y)
3
>>> degree(x**2+y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/polys/polytools.py", line 4414, in degree
    raise TypeError("generator needs to be specified for degree() for multivariate polynomials")
TypeError: generator needs to be specified for degree() for multivariate polynomials
```

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->
Fixes #13129

asmeurer commented:
I was going to suggest the error should also mention total_degree, but apparently we don't have a function for it yet (only a Poly method).

akash9712 commented:
I looked for a total_degree function too, but couldn't find it.
Also, should the function be changed so that in case of multiple generators being specified, a list of degrees with respect to the generators is returned?

asmeurer commented:
We already have `degree_list` for that functionality.

smichr commented:
Please start with a fresh version of master that has the modified version of the degree function.

smichr commented:
The generator is always supplied by the default argument (in master) so a more appropriate message might be

```
raise TypeError(filldedent('''
    a symbolic generator of interest is required for a multivariate
    polynomial, e.g. degree(x + y**2, gen=x) not degree(x + y**2, gen=0)
    '''))
```

akash9712 commented:
I am sorry I have to ask, but should multivariate expressions with a non zero constant generator raise an error?
For example: Should  ```degree(x**2+y**3,1)```  raise an error?
I am asking because the test_polytools contain tests regarding the same that return a value, and I can't understand the reasons.
https://github.com/sympy/sympy/blob/master/sympy/polys/tests/test_polytools.py#L1170

smichr commented:
In case of the Poly there is no ambiguity of specifying a nonzero generator since the Poly.gens give the generators as a tuple. If you don't have a Poly you don't know the generator order so raising an error forces the user to be unambiguous with the level of information they have. Someone might say, "I know that when I make the expression a poly that this gen will be number 2." To that we say, "Then feel free to use the Poly method or pass the Poly so it is unambiguous." At least that's how I see it.

akash9712 committed:
degree(multivariate): multivariate expression now needs specified generator

degree() function now needs a non constant generator to be specified for the
expression, if the expression is not specified as Poly. Tests were modified.

Example:>>> x,y=symbols('x,y')
>>> degree(x**2+y**3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/akash/sympy/sympy/polys/polytools.py", line 4432, in degree
    expression, e.g. degree(x + y**2, gen=x) not degree(x + y**2, gen=0)'''))
TypeError:
a symbolic generator of interest is required for a multivariate
expression, e.g. degree(x + y**2, gen=x) not degree(x + y**2, gen=0)
>>> degree(x**2+y**3,1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/akash/sympy/sympy/polys/polytools.py", line 4432, in degree
    expression, e.g. degree(x + y**2, gen=x) not degree(x + y**2, gen=0)'''))
TypeError:
a symbolic generator of interest is required for a multivariate
expression, e.g. degree(x + y**2, gen=x) not degree(x + y**2, gen=0)

akash9712 commented:
@smichr I understand, thank you. I have updated the pull request, and the I have resolved the merge conflicts. Please review.

akash9712 commented:
@asmeurer @smichr  Is the total_degree defined only as a Poly method, and not as a function on purpose? Should I try to implement it as a function?

smichr commented on sympy/polys/polytools.py:
```diff
@@ -4427,6 +4427,10 @@ def degree(f, gen=0):
         if gen not in p.gens:
             return S.Zero
```
your added code should be an elif clause:

```
if not sympify(gen, strict=True).is_Number:
    ...
elif not f.is_Poly and len(f.free_symbols) > 1:
    raise TypeError(filldedent('''
            A symbolic generator of interest is required when passing
            a multivariate expression like foo = %s, e.g. degree(foo, gen=%s)
            instead of degree(foo, gen=%s).
        ''' % (f, ordered(f.free_symbols).next(), gen)))
```

Note that the filldedent text comes on the next line so the error appears more neatly during interactive sessions. Also, an example specific to the user's passed multivariate expression is given instead of a generic expression.

e.g.
```
>>> f=x+y**2
>>> gen=1
>>> raise TypeError(filldedent('''
...             A symbolic generator of interest is required when passing
...             a multivariate expression like foo = %s, e.g. degree(foo, gen=%s
)
...             instead of degree(foo, gen=%s).
...         ''' % (f, ordered(f.free_symbols).next(), gen)))
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
TypeError:
A symbolic generator of interest is required when passing a
multivariate expression like foo = x + y**2, e.g. degree(foo, gen=x)
instead of degree(foo, gen=1).
```
If you use this, be sure that `ordered` is imported at the top of the file.

smichr commented on sympy/polys/polytools.py:
```diff
@@ -45,7 +45,7 @@
 )
 
-from sympy.utilities import group, sift, public
+from sympy.utilities import group, sift, public,filldedent
```
space after comma, please

smichr commented on sympy/polys/tests/test_polytools.py:
```diff
@@ -1166,10 +1166,10 @@ def test_Poly_degree():
-    assert degree(y**2+x**3, 1) == 2
     assert degree(pi) == 1
 
+    raises(TypeError, lambda: degree(y**2+x**3))
```
space around `+`, please

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into 13129_degree_Multivariate

akash9712 committed:
degree(multivariate): Modified if clause and edited error message

The if clause added previously is changed to an elif clause. The error
message is modified so as to give an example specific to the user's passed
multivariate expression, instead of a generic expression. White spaces are
added to test_util.py where needed.

smichr commented on sympy/polys/polytools.py:
```diff
@@ -4427,6 +4426,13 @@ def degree(f, gen=0):
         if gen not in p.gens:
             return S.Zero
 
+    elif len(f.free_symbols) > 1 and  sympify(gen).is_Number and not f.is_Poly:
```
This `elif` should read `elif not f.is_Poly and len(f.free_symbols) > 1:` because we already know that `sympify(gen).is_Number` is true because of the `if` clause. Also, there is no need to traverse the expression if we know that f is a Poly (and that, I assume is faster than finding the free symbols) -- put cheaper tests first.

akash9712 replied:
Thank you. Will take care of this next time.

smichr committed:
modify elif statement

smichr commented:
I made the edits directly and will commit if this passes if there is no objection.

smichr commented:
It's in!
